pkg/vsphere/compute/placement/random.go

```diff
@@ -32,28 +32,31 @@ func NewRandomHostPolicy() *RandomHostPolicy {
 }
 
 // CheckHost always returns false in a RandomHostPolicy.
-func (p *RandomHostPolicy) CheckHost(op trace.Operation, vm *vm.VirtualMachine) bool {
+func (p *RandomHostPolicy) CheckHost(op trace.Operation, sess *session.Session) bool {
 	return false
 }
 
 // RecommendHost recommends a random host on which to place a newly created VM.
-func (p *RandomHostPolicy) RecommendHost(op trace.Operation, vm *vm.VirtualMachine) (*object.HostSystem, error) {
-	r, err := vm.ResourcePool(op)
-	if err != nil {
-		return nil, err
+func (p *RandomHostPolicy) RecommendHost(op trace.Operation, sess *session.Session, hosts []*object.HostSystem) ([]*object.HostSystem, error) {
+	if hosts == nil {
+		rp := compute.NewResourcePool(op, sess, sess.Pool.Reference())
```

Comment:
Consider two different approaches:
1. NewRandomHostPolicy(sess *session) - creates a policy that's scoped for a given cluster. This avoids the issue you have with this call where the session supplied could be for a completely different cluster than the `[]hosts` were pulled from
2. NewRandomHostPolicy(hosts func() []*object.HostSystem) - takes a function that can be called to obtain the host list over which to place.

Reply:
Fixed - I embedded session into the policy struct. Keep in mind that the MetricsProvider also contains the session as well.